File: ios/Classes/models/RouteLegProgress.swift
import Foundation
import MapboxDirections
import MapboxNavigation
import MapboxCoreNavigation

public class MapBoxRouteLegProgress : Codable
{
    let leg : RouteLeg
    let distanceTraveled : Double
    let durationRemaining : Double
    let distanceRemaining : Double
    let fractionTraveled : Double
    let userHasArrivedAtWaypoint : Bool
    let stepIndex : Int
    let currentStep : MapBoxRouteStep 
    let shapeIndex : Int
    var priorStep : MapBoxRouteStep? = nil
    var upcomingStep : MapBoxRouteStep? = nil
    var followOnStep : MapBoxRouteStep? = nil
    // var currentStepProgress : MapBoxRouteStepProgress
    var currentSpeedLimit : Measurement<UnitSpeed>? = nil
    var remainingSteps : [MapBoxRouteStep] = []

    init(progress: RouteLegProgress) {
        leg = progress.leg
        distanceTraveled = progress.distanceTraveled
        durationRemaining = progress.durationRemaining
        distanceRemaining = progress.distanceRemaining
        fractionTraveled = progress.fractionTraveled
        userHasArrivedAtWaypoint = progress.userHasArrivedAtWaypoint
        stepIndex = progress.stepIndex
        remainingSteps = progress.remainingSteps.map { MapBoxRouteStep(step: $0) }
        priorStep = progress.priorStep != nil ? MapBoxRouteStep(step: progress.priorStep!) : nil
        currentStep = MapBoxRouteStep(step: progress.currentStep)
        upcomingStep = progress.upcomingStep != nil ? MapBoxRouteStep(step: progress.upcomingStep!) : nil
        followOnStep = progress.followOnStep != nil ? MapBoxRouteStep(step: progress.followOnStep!) : nil
        // currentStepProgress = MapBoxRouteStepProgress(stepProgress: progress.currentStepProgress)
        currentSpeedLimit = progress.currentSpeedLimit
        shapeIndex = progress.shapeIndex
    }
}

File: lib/src/models/helpers/route_leg_progress.dart
part of mapbox_navigation_flutter;

class RouteLegProgress {
  final RouteLeg? leg;
  final double distanceTraveled;
  final double fractionTraveled;
  final double durationRemaining;
  final double distanceRemaining;
  final int shapeIndex;
  final RouteStepProgress? routeStepProgress;
  final int currentSpeedLimit;

  RouteLegProgress._({
    required this.leg,
    required this.distanceRemaining,
    required this.currentSpeedLimit,
    required this.fractionTraveled,
    required this.distanceTraveled,
    required this.durationRemaining,
    required this.routeStepProgress,
    required this.shapeIndex,
  });

  static RouteLegProgress? fromAndroid(Map<String, dynamic>? json) => json !=
          null
      ? RouteLegProgress._(
          leg: RouteLeg.fromAndroid(json['routeLeg'] as Map<String, dynamic>),
          distanceRemaining: double.parse(json['distanceRemaining'].toString()),
          distanceTraveled: double.parse(json['distanceTraveled'].toString()),
          fractionTraveled: double.parse(json['fractionTraveled'].toString()),
          durationRemaining: double.parse(json['durationRemaining'].toString()),
          shapeIndex: int.parse(json['geometryIndex'].toString()),
          currentSpeedLimit: 0,
          routeStepProgress: RouteStepProgress.fromAndroid(
            json['currentStepProgress'] as Map<String, dynamic>?,
          ),
        )
      : null;

  static RouteLegProgress? fromIOS(Map<String, dynamic>? json) => json != null
      ? RouteLegProgress._(
          leg: RouteLeg.fromIOS(json['leg'] as Map<String, dynamic>),
          distanceRemaining: double.parse(json['distanceRemaining'].toString()),
          distanceTraveled: double.parse(json['distanceTraveled'].toString()),
          fractionTraveled: double.parse(json['fractionTraveled'].toString()),
          durationRemaining: double.parse(json['durationRemaining'].toString()),
          shapeIndex: int.parse(json['shapeIndex'].toString()),
          currentSpeedLimit:
              int.tryParse(json['currentSpeedLimit'].toString()) ?? -1,
          routeStepProgress: RouteStepProgress.fromIOS(
            json['currentStepProgress'] as Map<String, dynamic>?,
          ),
        )
      : null;
}
